### DISCUSSION TITLE
Cannot build libpsl on Windows

### DISCUSSION THREAD

**Discussion by zhcnyuyang** [curl/curl-for-win#65] [Q&A] [answered]
I am using Windows11 system and compiling libcurl on it. When running cmake it requires libssh2 and libpsl. I succesfully compiled libssh2 and imported it into cmake. But when compiling cmake I met a trouble. I tried all the ways to compile it in their offical documation but none of them succeeded. There is no articles online to introduce how to compile it on Windows. So when you were building libcurl on Windows, how did you solve this problem? And is there any libpsl binaries which has finished compiling and can be directly used? Thanks very much.

**Comment by zhcnyuyang** [comment]
I got libpsl source code from [here](https://github.com/rockdaboot/libpsl). And I tried the way of compiling in [README.MSVC.md](https://github.com/rockdaboot/libpsl/blob/master/README.MSVC.md). However, doesn't work.

**Comment by vszakats** [comment, marked as answer]
I solved it by building libpsl manually (without using its build-systems). You can see it in the file `libpsl.sh`.

In the curl-for-win official downloads there is a `libpsl.a` that is usable as-is (for static linking), but this is for mingw-w64 and not compatible with MSVC.

**Reply by WolfJMZ** [reply]
How can we get a MSVC static libpsl library?

**Comment by zhcnyuyang** [comment]
> I solved it by building libpsl manually (without using its build-systems). You can see it in the file `libpsl.sh`.
> 
> In the curl-for-win official downloads there is a `libpsl.a` that is usable as-is (for static linking), but this is for mingw-w64 and not compatible with MSVC.

libpsl.sh? Where?
After your replying, I still have some questions. First, what's the exact procedure of compile it on Mingw64? I generated the configure file according to its document but after I got the configure file, I got error when running ./configure command.
And how to use the .a file? Treat it to be the same as .lib file?

**Comment by vszakats** [comment]
In this very repository we're discussing this. → https://github.com/curl/curl-for-win/blob/main/libpsl.sh

It doesn't use `./configure`.

I suggest directing your questions to the libpsl project. In this project we're building curl.

**Comment by zhcnyuyang** [comment]
> I solved it by building libpsl manually (without using its build-systems). You can see it in the file `libpsl.sh`.
> 
> In the curl-for-win official downloads there is a `libpsl.a` that is usable as-is (for static linking), but this is for mingw-w64 and not compatible with MSVC.

I said "I generated the configure file according to its document but after I got the configure file" means I used mingw64 in msys2.

**Comment by vszakats** [comment]
Yeah, but this isn't the libpsl project, and as mentioned, we don't use autotools to build libpsl.
To get support for libpsl, please contact the libpsl project directly.

**Comment by zhcnyuyang** [comment]
> Yeah, but this isn't the libpsl project, and as mentioned, we don't use autotools to build libpsl. To get support for libpsl, please contact the libpsl project directly.

Curl for win? What's the project?

**Comment by vszakats** [comment]
This project is building curl for Windows and other operating systems.

**Comment by zhcnyuyang** [comment]
> This project is building curl for Windows and other operating systems.

What's the difference between it and curl?

**Comment by zhcnyuyang** [comment]
> This project is building curl for Windows and other operating systems.

And I entered project, didn't see any code and binaries. How to use this project? And what's the purpose of this project? Download code and compile them? Work with the code you download previously? I want to get a tutorial guide of it.

**Comment by vszakats** [comment]
No difference in context of libpsl.

**Comment by zhcnyuyang** [comment]
Too much new thing I've got puzzled...

**Comment by vszakats** [comment]
These are shell scripts to download and build curl and its dependencies. Most would not use this project directly but download the binaries it produced and run or link them to their projects (see <https://curl.se/windows/>).

This isn't actively supported, but you can also do your own builds by running the scripts yourself, locally, or via CI. See the CI entry points and `_build.sh` / `_build-runner.sh` for hints on how to.

**Comment by zhcnyuyang** [comment]
> These are shell scripts to download and build curl and its dependencies. Most would not use this project directly but download the binaries it produced and run or link them to their projects (see https://curl.se/windows).
> 
> This isn't actively supported, but you can also do your own builds by running the scripts yourself, locally, or via CI. See the CI entry points and `_build.sh` / `_build-runner.sh` for hints on how to.

CI? What's that?

**Comment by zhcnyuyang** [comment]
> These are shell scripts to download and build curl and its dependencies. Most would not use this project directly but download the binaries it produced and run or link them to their projects (see https://curl.se/windows).
> 
> This isn't actively supported, but you can also do your own builds by running the scripts yourself, locally, or via CI. See the CI entry points and `_build.sh` / `_build-runner.sh` for hints on how to.

Okay, I see.

**Comment by zhcnyuyang** [comment]
> These are shell scripts to download and build curl and its dependencies. Most would not use this project directly but download the binaries it produced and run or link them to their projects (see https://curl.se/windows).
> 
> This isn't actively supported, but you can also do your own builds by running the scripts yourself, locally, or via CI. See the CI entry points and `_build.sh` / `_build-runner.sh` for hints on how to.

And I have the last question is that in curl-for-win, to build the whole project, which one is the entrance? And What's my work after running the script?

**Reply by vszakats** [reply]
On a local machine `_build-runner.sh` is the best way to start. Look into it, edit it as needed and make sure that necessary toolchain is installed on the machine. If everything went well, the scripts will leave the distro tarball/zip on the disk.

Be prepared that this will take some effort, because as I said this isn't actively supported or "foolproof". I also haven't tested running these script on a real Windows machine (under MSYS2) for a long time.

The easier way is to fork the project and run it in GitHub Actions under your own account.

**Reply by zhcnyuyang** [reply]
What's included in the toolchain?

**Reply by vszakats** [reply]
It's the build tools (e.g. cmake), compiler components (llvm/gcc/binutils) and mingw-w64.
Look into the `_ci*.sh` files for a complete list.

**Comment by vszakats** [comment]
> CI? What's that?

AppVeyor CI, GitHub Actions: Continuous Integration hosts.

**Comment by tfloodeenjr** [comment]
How can we disable this functionality in curl, since it is clearly not ready for use with msvc?

**Reply by WolfJMZ** [reply]
Due to the vulnerability it's not recommended - building libpsl can be done on MSVC on windows and linked statically - if you need build script samples for MSVC on windows for building statically check out - https://gitlab.com/cplusplusdependencies

**Reply by WolfJMZ** [reply]
if you *really* want to disable the requirement - you can set (in cmake) -DCURL_USE_LIBPSL=NO but then it's strongly recommended to also disable cookies with -DCURL_DISABLE_COOKIES=ON

**Reply by vszakats** [reply]
libpsl's latest version builds fine, without dependencies, for Windows.
This project builds it manually with 3 lines of script, one of them needs
Python. It should be similar with MSVC, but I haven't tried it:
https://github.com/curl/curl-for-win/blob/43b175174882850be305a3b79e92152dcc081d00/libpsl.sh#L25-L36

**Reply by tfloodeenjr** [reply]
This also seems to work to disable it:

cmake.exe -D "CURL_USE_OPENSSL=ON" -D "CURL_USE_LIBPSL=OFF" -DCMAKE_INSTALL_PREFIX=C:\Build\curl\install C:\Build\\curl\curl-8.12.1 -G "Visual Studio 16 2019" -A x64

I will try the build script sample from https://gitlab.com/cplusplusdependencies as well.
If use the download shown in the provided "build.bat", libpsl-0.21.5.tar.gz, it seems to build the psl.lib library OK.

Thanks!

**Reply by tfloodeenjr** [reply]
The curl build now finds the header and library, but now will not link:

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.exe /ERRORREPORT:QUEUE /OUT:"C:\Build\curl\win64\lib\Release\libcurl.dll" /INCREMENTAL:NO /NOLOGO C:\Build\curl\deps\lib\libssl.lib C:\Build\curl\deps\lib\libcrypto.lib C:\Build\curl\deps\lib\zlib.lib wldap32.lib C:\Build\curl\deps\lib\psl.lib ws2_32.lib bcrypt.lib advapi32.lib crypt32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /DEF:"C:/Build/curl/curl-8.12.1/lib/libcurl.def" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:"C:/Build/curl/win64/lib/Release/libcurl.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/Build/curl/win64/lib/Release/libcurl_imp.lib" /MACHINE:X64  /machine:x64 /DLL libcurl_shared.dir\Release\libcurl.res

         cookie.obj : error LNK2019: unresolved external symbol __imp_psl_is_cookie_domain_acceptable referenced in function Curl_cookie_add [C:\Build\curl\win64\lib\libcurl_shared.vcxproj]
         psl.obj : error LNK2019: unresolved external symbol __imp_psl_free referenced in function Curl_psl_destroy [C:\Build\curl\\win64\lib\libcurl_shared.vcxproj]
         psl.obj : error LNK2019: unresolved external symbol __imp_psl_builtin referenced in function Curl_psl_use [C:\\Build\\curl\win64\lib\libcurl_shared.vcxproj]
         psl.obj : error LNK2019: unresolved external symbol __imp_psl_latest referenced in function Curl_psl_use [C:\Build\\curl\win64\lib\libcurl_shared.vcxproj]
         version.obj : error LNK2019: unresolved external symbol __imp_psl_check_version_number referenced in function curl_version [C:\Build\curl\win64\lib\libcurl_shared.vcxproj]
         C:\Build\curl\win64\lib\Release\libcurl.dll : fatal error LNK1120: 5 unresolved externals [C:\Build\\curl\win64\lib\libcurl_shared.vcxproj]

**Reply by WolfJMZ** [reply]
There's an additional C_FLAG needed because build.bat builds libpsl statically - if you look in the build.bat for curl in that same group of repos you'll see it adds `-DPSL_STATIC` to the `C_FLAGS` for building cURL.

https://gitlab.com/cplusplusdependencies/curl/-/blob/master/builds/msvc/build.bat?ref_type=heads

**Reply by WolfJMZ** [reply]
Again it's HIGHLY ill-advised to disable libpsl because it could allow for super-cookies which combined with other vulnerabilities could lead to leaks, information leaks, privacy issues, and other sign-in security issues.

**Reply by WolfJMZ** [reply]
As far as building libpsl - you can also use meson to build it with MSVC but I haven't tried that yet.

**Reply by tfloodeenjr** [reply]
Thanks, this cmake command line appears to do the trick linking with static libpsl 

cmake.exe -D "CURL_USE_OPENSSL=ON" -DCMAKE_CXX_FLAGS="/DPSL_STATIC" -DCMAKE_C_FLAGS="/DPSL_STATIC" -DCMAKE_INSTALL_PREFIX=C:\Build\curl\install C:\Build\curl\curl-8.12.1 -G "Visual Studio 16 2019" -A x64

I appreciate the assistance.

**Reply by vszakats** [reply]
> -DCMAKE_CXX_FLAGS

You can probably drop this option, the curl build doesn't use C++.
